src/v/cloud_roles/refresh_credentials.h

```diff
@@ -0,0 +1,184 @@
+
+        /// Calculates sleep duration given the number of seconds when the
+        /// credentials will expire. Keeps a small buffer to allow for network
+        /// calls
+        ss::lowres_clock::duration
+        calculate_sleep_duration(uint32_t expiry_sec) const;
+
+        /// When a retryable error is seen, increment retries and set a small
+        /// backoff before attempting to fetch credentials again. If the retries
+        /// go over the max limit, abort the operation by throwing an exception.
+        void increment_retries();
+
+        void reset_retries() {
+            vlog(
+              clrl_log.info, "resetting retry counter from {} to 0", _retries);
+            _retries = 0;
+        }
+
+        uint8_t retries() const { return _retries; }
+
+        const ss::sstring& api_host() const { return _api_host; }
+
+        uint16_t api_port() const { return _api_port; }
+
+        s3::aws_region_name region() const { return _region; }
+
+    private:
+        /// The hostname to query for credentials. Can be overridden using env
+        /// variable `RP_SI_CREDS_API_HOST`
+        ss::sstring _api_host;
+
+        /// The port to query for credentials. Can be overridden using env
+        /// variable `RP_SI_CREDS_API_PORT`
+        uint16_t _api_port;
+        s3::aws_region_name _region;
+        uint8_t _retries{0};
+
+        /// The duration to sleep before fetching credentials. Derived from
+        /// credentials expiry time adjusted with a buffer to account for
+        /// network calls.
+        std::optional<ss::lowres_clock::duration> _sleep_duration{std::nullopt};
+        ss::abort_source& _as;
+        retry_params _retry_params;
+    };
+
+    refresh_credentials(
+      std::unique_ptr<impl> impl,
+      ss::gate& gate,
+      ss::abort_source& as,
+      credentials_update_cb_t creds_update,
+      s3::aws_region_name region);
+
+    void start();
+
+    std::ostream& print(std::ostream& os) const { return _impl->print(os); }
+
+    ss::future<api_response> fetch_credentials() {
+        return _impl->fetch_credentials();
+    }
+
+private:
+    ss::future<> do_start();
+
+    /// Fetch the credentials from API using impl and trigger the callback
+    /// function with valid credentials. If an error was encountered when
+    /// fetching credentials, handle retry or abort based on the error.
+    ss::future<> fetch_and_update_credentials();
+
+    api_response_parse_result handle_response(api_response resp) {
+        return _impl->handle_response(std::move(resp));
+    }
+
+    ss::future<> sleep_until_expiry() const {
+        return _impl->sleep_until_expiry();
+    }
+
+private:
+    std::unique_ptr<impl> _impl;
+    ss::gate& _gate;
+    ss::abort_source& _as;
+    credentials_update_cb_t _credentials_update;
+    s3::aws_region_name _region;
+    auth_refresh_probe _probe;
+};
+
+std::ostream& operator<<(std::ostream& os, const refresh_credentials& rc);
+
+struct endpoint {
```

Comment:
look in `net` for `unresolved_addr`

Reply:
thanks for the pointer, I've removed this endpoint struct and used `net::unresolved_addr`, will explore completely removing the host+port combination from other places in `refresh_credentials` etc next.